distributed/shuffle/_arrow.py

```diff
@@ -85,3 +77,20 @@ def deserialize_table(buffer: bytes) -> pa.Table:
 
     with pa.ipc.open_stream(pa.py_buffer(buffer)) as reader:
         return reader.read_all()
+
+
+def read_from_disk(path: Path, meta: pd.DataFrame) -> tuple[Any, int]:
+    import pyarrow as pa
+
+    shards = []
+    schema = pa.Schema.from_pandas(meta, preserve_index=True)
+    with pa.OSFile(str(path), mode="rb") as f:
+        size = f.seek(0, whence=2)
+        f.seek(0)
+        while f.tell() < size:
+            sr = pa.RecordBatchStreamReader(f)
+            shard = sr.read_all()
+            arrs = [pa.concat_arrays(column.chunks) for column in shard.columns]
+            shard = pa.table(data=arrs, schema=schema)
+            shards.append(shard)
```

Comment:
By interleaving disk reads and deserialization, we reduced the size of the individual buffers that get created.